src/lib.rs

```diff
@@ -0,0 +1,186 @@
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![cfg_attr(feature="clippy", feature(plugin))]
+#![cfg_attr(feature="clippy", plugin(clippy))]
+
+extern crate libc;
+
+use libc::{pid_t, c_char, c_void, getpid, size_t};
+use std::ffi::CString;
+use std::path::PathBuf;
+use std::mem::transmute;
+
+// FFI stubs
+#[link(name = "traceme")]
+extern "C" {
+    fn traceme_start_tracer(conf: *const c_void) -> *const c_void;
+    fn traceme_stop_tracer(tr_ctx: *const c_void);
+}
+
+// Struct used to communicate a tracing configuration to C. Must stay in sync with the C code.
+#[repr(C)]
+struct TracerConf {
+    target_pid: pid_t,
+    trace_filename: *const c_char,
+    map_filename: *const c_char,
+    data_bufsize: size_t,
+    aux_bufsize: size_t,
+}
+
+/// Represents an instance of the Intel PT tracer.
+#[derive(Default)]
+pub struct Tracer {
+    /// Filename to store the trace to.
+    trace_filename: CString,
+    /// PID to trace.
+    target_pid: pid_t,
+    /// Data buffer size, in pages. Must be a power of 2.
+    data_bufsize: size_t,
+    /// Aux buffer size, in pages. Must be a power of 2.
+    aux_bufsize: size_t,
+    /// Opaque C pointer representing the tracer context.
+    tracer_ctx: Option<*const c_void>,
+}
+
+impl Tracer {
+    /// Create a new tracer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// Tracer.new().trace_filename("mytrace.ptt").data_bufsize(1024).target_pid(666);
+    /// ```
+    pub fn new() -> Self {
+        Tracer {
+            trace_filename: CString::new("traceme.ptt").unwrap(),
+            target_pid: unsafe { getpid() },
+            tracer_ctx: None,
+            data_bufsize: 64,
+            aux_bufsize: 1024,
+        }
+    }
+
+    /// Set where to write the trace.
+    ///
+    /// # Arguments
+    ///
+    /// * `filename` - Filename to use.
+    pub fn trace_filename(mut self, filename: &str) -> Self {
+        self.trace_filename = CString::new(filename).expect("bad trace filename");
+        self
+    }
+
+    /// Select which PID to trace.
+    ///
+    /// By default, the current PID is traced.
+    ///
+    /// # Arguments
+    ///
+    /// * `pid` - PID to trace.
+    pub fn target_pid(mut self, pid: pid_t) -> Self {
+        self.target_pid = pid;
+        self
+    }
+
+    /// Set the PT data buffer size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The size to use.
```

Comment:
Is there a suggested size or range of sizes? How do people know if they should set 8 or 10000000?

Reply:
It depends how many events you want to collect and with what latency.

A small buffer will require more frequent copy outs, a large buffer will consume more memory. We are going to have to experiment with these when the time comes.

Reply:
OK.